applications/GeoMechanicsApplication/custom_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_mass_and_damping.h

```diff
@@ -0,0 +1,634 @@
+
+                        rFirstDerivativeVector[id_z] = it->FastGetSolutionStepValue(VELOCITY_Z);
+                        rSecondDerivativeVector[id_z] = it->FastGetSolutionStepValue(ACCELERATION_Z);
+                    }
+                }
+            }
+        }
+    }
+
+
+    void BuildRHSNoDirichlet(
+        typename TSchemeType::Pointer pScheme,
+        ModelPart& rModelPart,
+        TSystemVectorType& rb)
+    {
+        KRATOS_TRY
+
+        //Getting the Elements
+        ElementsArrayType& r_elements = rModelPart.Elements();
+
+        //getting the array of the conditions
+        ConditionsArrayType& r_conditions = rModelPart.Conditions();
+
+        const ProcessInfo& r_current_process_info = rModelPart.GetProcessInfo();
+
+        //contributions to the system
+        LocalSystemVectorType rhs_contribution = LocalSystemVectorType(0);
+
+        //vector containing the localization in the system of the different
+        //terms
+        Element::EquationIdVectorType equation_ids;
+
+        // assemble all elements
+
+        const int nelements = static_cast<int>(r_elements.size());
+        #pragma omp parallel firstprivate(nelements, rhs_contribution, equation_ids)
+        {
+            #pragma omp for schedule(guided, 512) nowait
+            for (int i=0; i<nelements; i++) {
+                typename ElementsArrayType::iterator it = r_elements.begin() + i;
+                // If the element is active
+                if(it->IsActive()) {
+
+                    //calculate elemental Right Hand Side Contribution
+                    it->CalculateRightHandSide(rhs_contribution, r_current_process_info);
+                    it->EquationIdVector(equation_ids, r_current_process_info);
+
+                    //assemble the elemental contribution
+                    BaseType::AssembleRHS(rb, rhs_contribution, equation_ids);
+                }
+            }
+
+            rhs_contribution.resize(0, false);
+
+            // assemble all conditions
+            const int nconditions = static_cast<int>(r_conditions.size());
+            #pragma omp for schedule(guided, 512)
+            for (int i = 0; i<nconditions; i++) {
+                auto it = r_conditions.begin() + i;
+                // If the condition is active
+                if(it->IsActive()) {
+
+                    it->CalculateRightHandSide(rhs_contribution, r_current_process_info);
+                    it->EquationIdVector(equation_ids, r_current_process_info);
+
+                    //assemble the elemental contribution
+                    BaseType::AssembleRHS(rb, rhs_contribution, equation_ids);
+
+                }
+            }
+        }
+
+        TSystemVectorType first_derivative_vector;
+        TSystemVectorType second_derivative_vector;
+
+        GetFirstAndSecondDerivativeVector(first_derivative_vector, second_derivative_vector, rModelPart);
+
+        TSystemVectorType mass_contribution = prod(mMassMatrix, second_derivative_vector);
```

Comment:
@loumalouomega , is this the most efficient way of doing a dot product of a sparse matrix and vector?

Reply:
nope, the `TSparseSpace::Mult` method should be used I think

Reply:
Indeed, Ublas Spaces Mult must be used

Reply:
okay, thank you!